File: forthic-ts/src/forthic/ForthicError.ts
export interface ICodeLocation {
  screen_name: string;
  line: number;
  column: number;
  start_pos: number;
  end_pos: number;
}

export class ForthicError {
  error_key: string;
  title: string;
  description: string;
  location?: ICodeLocation;
  caught_error: any;

  constructor(
    error_key: string,
    title: string,
    description: string,
    location = null,
  ) {
    this.error_key = error_key;
    this.title = title;
    this.description = description;
    this.location = location;
    console.log("ForthixError", { error_key, title, description, location });
    this.caught_error = null;
  }

  set_caught_error(error: any) {
    this.caught_error = error;
  }

  get_title() {
    return this.title;
  }

  get_description() {
    return `${this.description}`;
  }

  get_error_stack() {
    const MAX_DEPTH = 100;

    let cur_error = this;
    const result = [cur_error];
    for (let i = 0; i < MAX_DEPTH; i++) {
      if (cur_error.caught_error) {
        result.push(cur_error.caught_error);
        cur_error = cur_error.caught_error;
      } else {
        break;
      }
    }
    return result.reverse();
  }
}

File: forthic-rb/lib/forthic/forthic_error.rb
# frozen_string_literal: true

module Forthic
  class ForthicError < StandardError
    attr_accessor :error_key, :title, :description, :location, :caught_error

    # @param [String] error_key
    # @param [String] title
    # @param [String] description
    # @param [CodeLocation, nil] location
    def initialize(error_key, title, description, location = nil)
      @error_key = error_key
      @title = title
      @description = description
      @location = location
      @caught_error = nil
    end

    # @param [ForthicError] error
    def set_caught_error(error)
      @caught_error = error
    end

    # @return [String]
    def get_title
      @title
    end

    # @return [String]
    def get_description
      @description
    end

    # @return [Array<ForthicError>]
    def get_error_stack
      max_depth = 100
      cur_error = self
      result = [cur_error]

      max_depth.times do
        break unless cur_error.caught_error

        result << cur_error.caught_error
        cur_error = cur_error.caught_error
      end

      result.reverse
    end
  end
end
